### PR TITLE
optimisations

### PR DESCRIPTION
**Pull request by andresailer**
See also AIDAsoft/aidaTT#19

BEGINRELEASENOTES
- Performance optimisations: avoiding temporaries, divisions, sqrts
- protect against too many tracks

ENDRELEASENOTES

### CODE REVIEW

File: src/KDTrack.C [hunk 1]
```diff
@@ -96,15 +97,15 @@ double KDTrack::calculateChi2SZ(TH2F* histo, bool debug) {
...
-  double sPhi0 = sin(phi0);
-  double ratio = (y0 - b) / (x0 - a);
-  double errorPhi0 =
-      (1. / ((y0 - b) * (1 + (ratio * ratio)))) * sqrt((((errorx0 * errorx0 + da * da) * ratio * ratio * ratio * ratio) +
-                                                        (errory0 * errory0 + db * db) * ratio * ratio));
+  const double cPhi0              = cos(phi0);
+  const double sPhi0              = sin(phi0);
+  const double ratio              = (y0 - b) / (x0 - a);
+  const double ratio2             = ratio * ratio;
+  const double errorPhi0Denomitor = (1. / ((y0 - b) * (1 + (ratio2))));
```
Comment (eleogran) [inline src/KDTrack.C]: DenomiNAtor I guess

File: src/ConformalTracking.cc [hunk 2]
```diff
@@ -621,12 +627,28 @@ void ConformalTracking::processEvent(LCEvent* evt) {
...
+    caughtException = false;
+    try {
+      Parameters displacedParameters(m_maxCellAngle * factor, m_maxCellAngleRZ * factor, m_chi2cut * factor, 5, 0.015, false,
+                                     true);
+      streamlog_out(DEBUG8) << "Building new tracks displaced: " << factor << std::endl;
+      buildNewTracks(conformalTracks, kdClusters, nearestNeighbours, displacedParameters, true, false);
+    } catch (TooManyTracksException& e) {
+      streamlog_out(MESSAGE) << "caught too many tracks, tightening parameters" << std::endl;
+      caughtException = true;
+      factor -= 1;
```
Comment (eleogran) [inline src/ConformalTracking.cc]: how much is the reduction in number of tracks from factor = 10 to factor = 9? Isn't the -= 1 bin too fine?

File: src/KDTrack.C [hunk 3]
```diff
@@ -191,20 +195,17 @@ double KDTrack::calculateChi2SZ(TH2F* histo, bool debug) {
...
 
     // Debug info
-    if (debug)
+    if (debug) {
       streamlog_out(DEBUG5) << "- hit " << hit << " has residualS = " << residualS << ", with error dz = " << errorZ
                             << ", error ds = " << sqrt(errorS2) << " and total error = " << sqrt(ds2) << std::endl;
-
-    double debugError2 = sqrt(deltaPhi * deltaPhi * errorRadC * errorRadC + errorDeltaPhi * errorDeltaPhi * radC * radC);
-    if (debug)
+      double debugError2 = sqrt(deltaPhi * deltaPhi * errorRadC * errorRadC + errorDeltaPhiSquared * radC * radC);
```
Comment (eleogran) [inline src/KDTrack.C]: minor: since it is a sqrt, why is this named Error2?

Comment (andresailer) [reply src/KDTrack.C]: the name was there before, but I will rename it

File: src/ConformalTracking.cc (hunk 2 above)
Comment (andresailer) [reply src/ConformalTracking.cc]: I will add a flag to disable the tightening of parameters, and if disabled directly skip the event.

File: src/KDTrack.C (hunk 1 above)
Comment (andresailer) [reply src/KDTrack.C]: of course

**Pull request merged**